Stephen Kay <br>
CSCI-3287 Course Project <br>
Gaming the System <br>

In [2]:
%load_ext sql
%matplotlib inline

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [3]:
import matplotlib.pyplot as plt
import zipfile
import pandas as pd
from sqlalchemy import create_engine
import sqlalchemy.sql

In [4]:
import getpass
# passwd = getpass.getpass()
passwd = 'M6f9NYDl6FgaBMd1jyLBxpAYVXHcBXl3'
dbuser='qktewplm'
eng = create_engine('postgresql://{0}:{1}@otto.db.elephantsql.com:5432/qktewplm'.format(dbuser, passwd))
con = eng.connect()

In [5]:
con.execute("SELECT postgis_full_version()").fetchall()

[('POSTGIS="2.5.1 r17027" [EXTENSION] PGSQL="110" GEOS="3.6.2-CAPI-1.10.2 4d2925d6" PROJ="Rel. 4.9.3, 15 August 2016" GDAL="GDAL 2.2.3, released 2017/11/20" LIBXML="2.9.4" LIBJSON="0.12.1" LIBPROTOBUF="1.2.1" RASTER',)]

In [6]:
rs = con.execute('''
DROP TABLE IF EXISTS nba_schedule;
CREATE TABLE nba_schedule
	(
	  id numeric,
	  road_team character varying(50),
	  home_team character varying(50),
	  total_game_minutes numeric,
	  road_team_final_score numeric,
	  home_team_final_score numeric,
      road_team_rest character varying(50),
      home_team_rest character varying(50)
	);
''')

In [7]:
con.execute("SELECT * from nba_schedule LIMIT 1").fetchall()

[]

## Ok now let's use the wonderful Pandas to read in our CSV and convert it into a dictionary! ##

In [8]:
nba_data = pd.read_csv("2017-2018_NBA_Historical_Schedule.csv")

In [9]:
nba_data.head()

id  road_team     home_team  total_game_minutes  \
0  21700001     Boston     Cleveland               240.0   
1  21700002    Houston  Golden State               240.0   
2  21700003  Charlotte       Detroit               240.0   
3  21700004   Brooklyn       Indiana               240.0   
4  21700005      Miami       Orlando               240.0   

   road_team_final_score  home_team_final_score road_team_rest home_team_rest  
0                     99                    102             3+             3+  
1                    122                    121             3+             3+  
2                     90                    102             3+             3+  
3                    131                    140             3+             3+  
4                    109                    116             3+             3+

In [10]:
nba_dict = nba_data.to_dict('records')

In [11]:
nba_dict[0]

{'id': 21700001,
 'road_team': 'Boston',
 'home_team': 'Cleveland',
 'total_game_minutes': 240.0,
 'road_team_final_score': 99,
 'home_team_final_score': 102,
 'road_team_rest': '3+',
 'home_team_rest': '3+'}

In [12]:
for row in nba_dict:
    cmd = sqlalchemy.sql.text('''INSERT INTO nba_schedule(id, road_team, home_team, total_game_minutes,
    road_team_final_score, home_team_final_score, road_team_rest, home_team_rest)\
    VALUES ( :id, :road_team, :home_team, :total_game_minutes, :road_team_final_score,
    :home_team_final_score, :road_team_rest, :home_team_rest) ''' )
    con.execute(cmd, row)

In [13]:
con.execute("SELECT * from nba_schedule order by id asc LIMIT 1").fetchall()

[(Decimal('21700001'), 'Boston', 'Cleveland', Decimal('240.0'), Decimal('99'), Decimal('102'), '3+', '3+')]

In [14]:
nba_dict[0]

{'id': 21700001,
 'road_team': 'Boston',
 'home_team': 'Cleveland',
 'total_game_minutes': 240.0,
 'road_team_final_score': 99,
 'home_team_final_score': 102,
 'road_team_rest': '3+',
 'home_team_rest': '3+'}

In [15]:
con.execute("SELECT * from nba_schedule order by id desc LIMIT 1").fetchall()

[(Decimal('41700404'), 'Golden State', 'Cleveland', Decimal('240.0'), Decimal('108'), Decimal('85'), '1', '1')]

In [16]:
data_list_len = len(nba_data)
nba_dict[data_list_len-1]

{'id': 41700404,
 'road_team': 'Golden State',
 'home_team': 'Cleveland',
 'total_game_minutes': 240.0,
 'road_team_final_score': 108,
 'home_team_final_score': 85,
 'road_team_rest': '1',
 'home_team_rest': '1'}

The above SQL and dictionary statements ensure that my database has all of the rows by checking the database first and last entries against the dictionaries' view of the first and last entries.

In [17]:
import geopandas

In [18]:
nba_cities_data = pd.read_csv("NBA-Cities-Lat-Long.csv")

In [19]:
# from shapely.geometry import Point

# nba_cities_data['geometry'] = list(zip(nba_cities_data.lon, nba_cities_data.lat))
# nba_cities_data['geometry'] = nba_cities_data['geometry'].apply(Point)

In [20]:
rs = con.execute('''
DROP TABLE IF EXISTS nba_cities;
CREATE TABLE nba_cities
	(
	  city character varying(50),
      lat numeric,
      lon numeric
	);
''')

In [21]:
for row in nba_cities_dict:
    row['wkt'] = row['geometry'].wkt
    cmd = sqlalchemy.sql.text('''INSERT INTO nba_cities(city, lat, lon)\
    VALUES (:city, :lat, :lon) ''' )
    con.execute(cmd, row)

NameError: name 'nba_cities_dict' is not defined

In [ ]:
con.execute("SELECT * from nba_cities LIMIT 1").fetchall()

Ok this looks good! Now let's try and do some joins on our tables!

In [ ]:
con.execute('''ALTER TABLE nba_cities DROP COLUMN IF EXISTS geog;
               ALTER TABLE nba_cities ADD COLUMN geog GEOGRAPHY(POINT,4326)''')

In [ ]:
con.execute("UPDATE nba_cities SET geog = ST_SetSRID(ST_MakePoint(lon,lat),4326);")

In [ ]:
con.execute("SELECT * from nba_cities LIMIT 1").fetchall()

In [ ]:
game_total = con.execute("SELECT COUNT(*) from nba_schedule").fetchall()
print("There were, ", game_total, "total games in the 2017-2018 NBA season")

## Ok now that we have 2 initial table set up let's run some simple queries to get a handle on what we can begin to glean from the data ##

### <center>Some possible queries of interest are</center> ###

<center><li> What was the average score across all games for the road and home teams in the 2017-2018 season?</li><center>

<center><li> How many games in the 2017-2018 NBA season did road teams win when they were on a back-to-back or a 3IN4 and the home team had at least 1 or more nights of rest?</li><center>

<center><li> How many games in the 2017-2018 NBA season did home teams win when they had 1 or more nights of rest and the road team was on a B2B or a 3IN4, what?</li><center>

In [ ]:
con.execute('''
SELECT AVG(home_team_final_score)
    FROM nba_schedule
LIMIT 5;
''').fetchall()

In [ ]:
con.execute('''
SELECT AVG(road_team_final_score)
    FROM nba_schedule
LIMIT 5;
''').fetchall()

In [ ]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a
### back-to-back and the home team had 1 night of rest
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score > home_team_final_score)
LIMIT 5;
''').fetchall()

In [ ]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score > home_team_final_score)
''').fetchall()))

In [ ]:
### How many games in the 2017-2018 NBA season did home teams win when they had 1 night
### of rest and the road team was on a B2B
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score < home_team_final_score)
LIMIT 5;
''').fetchall()

In [ ]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '1'
    AND (home_team_final_score > road_team_final_score)
''').fetchall()))

In [ ]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a
### back-to-back and the home team had 2 nights of rest
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score > home_team_final_score)
LIMIT 5;
''').fetchall()

In [ ]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score > home_team_final_score)
''').fetchall()))

In [ ]:
### How many games in the 2017-2018 NBA season did home teams win when they had 2 nighs
### of rest and the road team was on a B2B
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score < home_team_final_score)
LIMIT 5;
''').fetchall()

In [ ]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score < home_team_final_score)
''').fetchall()))

In [ ]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a
### 3IN4 and the home team had 1 nights of rest
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4'
    AND home_team_rest = '1'
    AND (road_team_final_score > home_team_final_score)
LIMIT 5;
''').fetchall()

In [ ]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4'
    AND home_team_rest = '1'
    AND (road_team_final_score > home_team_final_score)
''').fetchall()))

In [ ]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a
### 3IN4 and the home team had 1 nights of rest
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4'
    AND home_team_rest = '2'
    AND (road_team_final_score > home_team_final_score)
LIMIT 5;
''').fetchall()

In [ ]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4'
    AND home_team_rest = '2'
    AND (road_team_final_score > home_team_final_score)
''').fetchall()))

In [ ]:
### How many games in the 2017-2018 NBA season did home teams win when they had 1 night of
### rest and the road team was on a B2B
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score < home_team_final_score)
LIMIT 5;
''').fetchall()

In [ ]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4'
    AND home_team_rest = '1'
    AND (road_team_final_score < home_team_final_score)
''').fetchall()))

In [ ]:
### How many games in the 2017-2018 NBA season did home teams win when they had 2 nighs of
### rest and the road team was on a 3IN4
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4'
    AND home_team_rest = '2'
    AND (road_team_final_score < home_team_final_score)
LIMIT 5;
''').fetchall()

In [ ]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4'
    AND home_team_rest = '2'
    AND (road_team_final_score < home_team_final_score)
''').fetchall()))

In [ ]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a 
### 3IN4-B2B and the road team had at least one night of rest
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score > home_team_final_score)
LIMIT 5;
''').fetchall()

In [ ]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score > home_team_final_score)
''').fetchall()))

In [ ]:
### How many games in the 2017-2018 NBA season did home teams win when they had
### at leat 1 night of rest and the road team was on a 3IN4 and B2B
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score < home_team_final_score)
LIMIT 5;
''').fetchall()

In [ ]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score < home_team_final_score)
''').fetchall()))

In [ ]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a 
### 3IN4-B2B and the road team had 2 nights of rest
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score > home_team_final_score)
LIMIT 5;
''').fetchall()

In [ ]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score > home_team_final_score)
''').fetchall()))

In [ ]:
### How many games in the 2017-2018 NBA season did home teams win when they had
### at 2 nights of rest and the road team was on a 3IN4 and B2B
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score < home_team_final_score)
LIMIT 5;
''').fetchall()

In [ ]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score < home_team_final_score)
''').fetchall()))

# <center>Woof, I definitely have query fatigue from that, let's summarize it in a table</center> #

<table>
    <tr> <th> Win Total </th>
         <th> Road Team Rest </th>
         <th> Home Team Rest </th>
         <th> Win Total </th>
    </tr>
    <tr> <th> 15 </th>
         <th> B2B </th>
         <th> 1 </th>
         <th> 31 </th>
    </tr>
    <tr> <th> 5 </th>
         <th> B2B </th>
         <th> 2 </th>
         <th> 7 </th>
    </tr>
    <tr> <th> 26 </th>
         <th> 3IN4 </th>
         <th> 1 </th>
         <th> 37 </th>
    </tr>
    <tr> <th> 9 </th>
         <th> 3IN4 </th>
         <th> 2 </th>
         <th> 11 </th>
    </tr>
    <tr> <th> 36 </th>
         <th> 3IN4-B2B </th>
         <th> 1 </th>
         <th> 58 </th>
    </tr>
    <tr> <th> 11 </th>
         <th> 3IN4-B2B </th>
         <th> 2 </th>
         <th> 26 </th>
    </tr>
    <tr> <th> Sum Total of Road Wins </th>
         <th> ... </th>
         <th> ... </th>
         <th> Sum Total of Home Wins </th>
    </tr>
    <tr> <th> 102 </th>
         <th> ... </th>
         <th> ... </th>
         <th> 170 </th>
    </tr>
</table>

In summary when the home team is in a rest advantagous situation they win 170/272 or 62.5% of the time while when the road team is in a rest disadventageous situation they win 102/272 37.5% of the time

However these situations only make up for 272/1312 or 20% of all NBA games in a given season. Let's use geospacing queries to see if the travel distance makes any difference for the score of a NBA game

If we look at our nba schedule table we'll see that it would be really nice to have the latitude and longitude in that data to be able to run some geo queries. Fortunately our NBA cities table has this information we just need to JOIN the two tables in order to be able to run some more insightful geoqueries

In [ ]:
nba_data.head(1)

In [ ]:
nba_cities_dict[0]

Let's create a join on these two table so we can get the geometry points for the home team and the road team

In [ ]:
con.execute('''
UPDATE nba_schedule a
SET road_team_geog = b.geog
FROM nba_cities b
WHERE a.road_team = b.city
''')

In [ ]:
con.execute('''
SELECT * FROM nba_schedule LIMIT 5;
''').fetchall()

In [ ]:
con.execute('''
ALTER TABLE nba_schedule add road_team_geog GEOGRAPHY
''')

In [ ]:
con.execute('''
UPDATE nba_schedule 
SET road_team_geog = (SELECT nba_cities.geog 
    FROM nba_cities
    WHERE nba_cities.city = nba_schedule.road_team
    );
''')

In [ ]:
con.execute('''
SELECT * FROM nba_schedule LIMIT 5;
''').fetchall()

In [ ]:
con.execute('''
ALTER TABLE nba_schedule add home_team_geog GEOGRAPHY
''')

In [ ]:
con.execute('''
UPDATE nba_schedule 
SET home_team_geog = (SELECT nba_cities.geog 
    FROM nba_cities
    WHERE nba_cities.city = nba_schedule.home_team
    );
''')

In [ ]:
con.execute('''
SELECT * FROM nba_schedule LIMIT 50;
''').fetchall()